process/track/blockProcessor.go

```diff
@@ -233,31 +241,83 @@ func (bp *blockProcessor) requestHeadersIfNeeded(
 		return
 	}
 
-	nbSortedHeaders := len(sortedHeaders)
-	if nbSortedHeaders == 0 {
+	shouldRequestHeaders := false
+
+	defer func() {
+		if !shouldRequestHeaders {
+			bp.requestHeadersIfNothingNewIsReceived(lastNotarizedHeader, sortedHeaders, longestChainHeaders)
+		}
+	}()
+
+	numSortedHeaders := len(sortedHeaders)
+	if numSortedHeaders == 0 {
 		return
 	}
 
-	highestNonceReceived := sortedHeaders[nbSortedHeaders-1].GetNonce()
+	highestNonceReceived := sortedHeaders[numSortedHeaders-1].GetNonce()
 	highestNonceInLongestChain := lastNotarizedHeader.GetNonce()
-	nbLongestChainHeaders := len(longestChainHeaders)
-	if nbLongestChainHeaders > 0 {
-		highestNonceInLongestChain = longestChainHeaders[nbLongestChainHeaders-1].GetNonce()
+	numLongestChainHeaders := len(longestChainHeaders)
+	if numLongestChainHeaders > 0 {
+		highestNonceInLongestChain = longestChainHeaders[numLongestChainHeaders-1].GetNonce()
 	}
 
-	shouldRequestHeaders := highestNonceReceived > highestNonceInLongestChain+bp.blockFinality && nbLongestChainHeaders == 0
+	shouldRequestHeaders = highestNonceReceived > highestNonceInLongestChain+bp.blockFinality && numLongestChainHeaders == 0
 	if !shouldRequestHeaders {
 		return
 	}
 
 	log.Debug("requestHeadersIfNeeded",
 		"shard", lastNotarizedHeader.GetShardID(),
 		"last notarized nonce", lastNotarizedHeader.GetNonce(),
+		"numSortedHeaders", numSortedHeaders,
+		"numLongestChainHeaders", numLongestChainHeaders,
 		"highest nonce received", highestNonceReceived,
 		"highest nonce in longest chain", highestNonceInLongestChain)
 
-	shardID := lastNotarizedHeader.GetShardID()
-	fromNonce := highestNonceInLongestChain + 1
+	bp.requestHeaders(lastNotarizedHeader.GetShardID(), highestNonceInLongestChain+1)
+}
+
+func (bp *blockProcessor) requestHeadersIfNothingNewIsReceived(
+	lastNotarizedHeader data.HeaderHandler,
+	sortedHeaders []data.HeaderHandler,
+	longestChainHeaders []data.HeaderHandler,
+) {
+	if check.IfNil(lastNotarizedHeader) {
+		return
+	}
+
+	header := lastNotarizedHeader
+	numLongestChainHeaders := len(longestChainHeaders)
+	if numLongestChainHeaders > 0 {
+		header = longestChainHeaders[numLongestChainHeaders-1]
```

Comment:
this can never be reached as how the code is written now

Reply:
Could be, but anyway I have refactored this part